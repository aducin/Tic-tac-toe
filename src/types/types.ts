interface BoxBase {
    icon: string;
    iconWinning: string;
    value: string;
};

export interface Box extends BoxBase {
    assigned: boolean;
    winning: boolean;
};

export type BoxesProps = {
    boxes: Box[];
    finished: boolean;
    onBoxClicked: (id: number) => void;
    started: boolean;
};

export type BoxItemProps = {
    data: {
        id: number;
        item: Box,
    };
    onBoxClicked: (id: number) => void;
}

export type ButtonsProps = {
    onButtonClicked: (button: string) => void;
    started: boolean;
};

export type DetailsProps = {
    activePlayer: string;
    turnTimer: number;
    value: string;
}

export type GameComponentState = {
    gameTimer: number;
    message: string;
    success: boolean;
    turnTimer: number;
    warning: boolean;
};

export type GameHistory = {
    duration: number;
    finishedTime: string;
    moves: number;
    winner: string;
};

export interface GameType extends BoxBase {
    id: number;
};

export type HistoryAction = {
    duration: number;
    finishedTime: Date;
}

export type IsPossibleToWin = {
    first: number;
    second: number
};

export type MatchedRows = {
    first: number;
    second: number;
};

export type MessageProps = {
    message: string;
    onRemoveAlert?: () => void;
    success: boolean;
    warning: boolean;
};

type Player = {
    active: boolean;
    name: string;
}

export type PlayerAction = {
    property: string;
    value: string;
}

export type PlayersProps = {
    onInputChanged: (property: string, name: string) => void;
    player1: string;
    player2: string;
}

export type ReducerAction = {
    payload: {
        boxes?: Box[];
        player?: PlayerAction;
        scoresList?: HistoryAction;
    }
    type: string;
};

export interface ScoresProps {
    scores: GameHistory[] | [];
}

export interface StoreState extends ScoresProps {
    boxes: Box[];
    finished: boolean;
    moves: number;
    next: GameType;
    player1: Player;
    player2: Player;
    started: boolean;
};
