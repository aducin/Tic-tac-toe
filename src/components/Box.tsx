import React from 'react';
import { BoxItemProps } from '../interfaces/interfaces';
import { PLAYER_FIRST_TYPE } from '../constants/constants';
import ImageO from '../assets/images/o_normal.png';
import ImageX from '../assets/images/x_normal.png';
import ImageOWinning from '../assets/images/o_winning.png';
import ImageXWinning from '../assets/images/x_winning.png';
import styles from '../assets/styles/box.module.scss';

const icons = [ImageO, ImageX];
const iconsWinning = [ImageOWinning, ImageXWinning];

const BoxItem: React.FC<BoxItemProps> = (props: BoxItemProps) => {
    const { id, item } = props.data;
    let content;
    let iconFile;

    if (props.data.item.value) {
        const firstPlayerActive = props.data.item.value === PLAYER_FIRST_TYPE;
        
        if (item.winning) {
            iconFile = firstPlayerActive ? iconsWinning[0] : iconsWinning[1];
        } else {
            iconFile = firstPlayerActive ? icons[0] : icons[1];
        }

        if (item.assigned) {
            content = <img src={iconFile} className={styles.image} alt=""></img>;
        }
    }
    
    return (
        <div
            onClick={() => props.onBoxClicked(id)}
            className={styles.item}
        >{content}</div>
    );
}

export default React.memo(BoxItem);
