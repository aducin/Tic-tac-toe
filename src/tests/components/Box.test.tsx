import React from "react";
import TestRenderer from 'react-test-renderer';
import { shallow } from "enzyme";
import { ShallowWrapper } from 'enzyme';
import Box from '../../components/Box';
import { BoxItemProps } from '../../types/types';

const boxItemProps: BoxItemProps = {
    data: {
        id: 0,
        item: {
            assigned: true,
            icon: 'ImageO',
            value: 'O',
            winning: false,
            iconWinning: 'ImageOWinning'
        }
    },
    onBoxClicked: () => {}
}

const component = (
    <Box
        data={boxItemProps.data}
        onBoxClicked={boxItemProps.onBoxClicked}
    />
);

const testData = TestRenderer.create(component);

describe('Box component snapshot', () => {
    it('Box component renders correctly', async () => {
        expect(testData.toJSON()).toMatchSnapshot();
    });
});

describe('Box component tests', () => {
    let wrapper: ShallowWrapper;
    beforeEach(() => {
        wrapper = shallow(component);
    });

    it('there is an image when item assigned', () => {
        const el = wrapper.find('img');
        expect(el).toBeTruthy;
        expect(el.prop("src")).toEqual('o_normal.png');
    });

    it('there is no img tag when an item not assigned', () => {
        wrapper.setProps({ assigned: false });
        expect(wrapper.find('img')).toBeFalsy;
    });
});  
