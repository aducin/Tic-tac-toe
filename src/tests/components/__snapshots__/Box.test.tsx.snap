// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box component snapshot Box component renders correctly 1`] = `
<div
  className="item"
  onClick={[Function]}
>
  <img
    alt=""
    className="image"
    src="o_normal.png"
  />
</div>
`;
